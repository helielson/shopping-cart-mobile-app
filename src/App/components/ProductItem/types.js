/**
 * Types definition used on ProductItem Component
 *
 * @format
 * @flow
 */
import type { itemType } from '../../flow/types';

// PropTypes for ProductItem Component
export type ProductItemProps = {
  navigate: (item: itemType) => void,
  item: itemType,
  addOne: (item: itemType) => mixed,
};
