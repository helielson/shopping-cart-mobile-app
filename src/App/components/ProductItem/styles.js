import colors from '../../config/colors';

const shadowStyle = {
  shadowOpacity: 0.35,
  shadowOffset: {
    width: 0,
    height: 1,
  },
  shadowColor: '#000',
  shadowRadius: 2,
  elevation: 3,
};

export default {
  container: {
    alignItems: 'center',
    marginVertical: 10,
    marginHorizontal: 0,
    backgroundColor: colors.backgroundLight,
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    ...shadowStyle,
  },
  picture: {
    height: 110,
    flex: 2,
    backgroundColor: colors.background,
    borderWidth: 2,
    borderColor: colors.background,
  },
  rightContainer: {
    flex: 3,
    flexDirection: 'column',
    padding: 10,
    justifyContent: 'space-between',
  },
  priceContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around',
  },
  productTitle: {
    flex: 1,
    color: '#333',
    fontWeight: 'bold',
    fontSize: 15,
    textAlign: 'center',
  },
  productPrice: {
    flex: 1,
    color: colors.primary,
    fontSize: 16,
  },
  addButtonContainer: {
    backgroundColor: colors.buttonColor,
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 5,
    paddingHorizontal: 8,
    borderRadius: 5,
  },
  addButtonText: {
    color: colors.backgroundLight,
  },
};
