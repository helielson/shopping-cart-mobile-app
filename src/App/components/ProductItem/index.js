/**
 * @format
 * @flow
 */
import React from 'react';
import { connect } from 'react-redux';
import { View, Text, Image, TouchableOpacity } from 'react-native';
import type { ProductItemProps } from './types';
import { addOneItem } from '../../store/actions';
import Button from '../Button';
import formatCurrency from '../../util';
import styles from './styles';

/**
 * ProductItem component that allows user to add item to Cart
 *
 * @param {ProductItemProps} {
 *   item,
 *   navigate,
 *   addOne
 * }
 * @returns <ProductItem />
 */
const ProductItem = (props: ProductItemProps) => {
  const { navigate, item, addOne } = props;
  const {
    item: { picture, price, title },
  } = props;
  return (
    <TouchableOpacity
      activeOpacity={0.6}
      onPress={() => navigate(props.item)}
      style={styles.container}
    >
      <Image style={styles.picture} source={{ uri: picture }} />
      <View style={styles.rightContainer}>
        <Text style={styles.productTitle}>{title}</Text>
        <View style={styles.priceContainer}>
          <Text style={styles.productPrice}>{formatCurrency(price)}</Text>
          <Button
            withIcon
            iconName="add-shopping-cart"
            text="Add to Cart"
            onPress={() => {
              addOne(item);
            }}
          />
        </View>
      </View>
    </TouchableOpacity>
  );
};

// Map Redux Actions Dispatchers to Component Props
const mapDispatchToProps = dispatch => ({
  addOne: item => dispatch(addOneItem(item)),
});

// Export Stateless Component not connected to redux store (To be used on Tests)
export { ProductItem };

// Connect component with store and export it as default
export default connect(
  null,
  mapDispatchToProps
)(ProductItem);
