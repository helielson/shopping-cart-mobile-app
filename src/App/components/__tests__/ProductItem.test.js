/**
 * @flow
 * @format
 * */
import React from 'react';
import { shallow } from 'enzyme';

import { ProductItem } from '../ProductItem';

const item = {
  id: 'a',
  description:
    'Análise Sensorial: \n Análise Visual: cor rosa com reflexos alaranjados. Análise Olfativa: Frutas vermelhas (morango) Análise Gustativa: seco, corpo médio, excelente acidez, taninos de boa qualidade, com retro-olfato de frutas.',
  price: 200,
  available: 20,
  picture: 'https://cdn.awsli.com.br/600x700/365/365010/produto/13341816/fe9a59131b.jpg',
  title: 'Vinho Nederburg Rosé',
};

describe('<ProductItem />', () => {
  test('should render properly', () => {
    const wrapper = shallow(<ProductItem item={item} addOne={() => {}} navigate={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });
});
