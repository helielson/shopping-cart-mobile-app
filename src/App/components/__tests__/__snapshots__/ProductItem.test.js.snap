// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> should render properly 1`] = `
<TouchableOpacity
  activeOpacity={0.6}
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(245, 245, 246, 1.00)",
      "elevation": 3,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginHorizontal": 0,
      "marginVertical": 10,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.35,
      "shadowRadius": 2,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://cdn.awsli.com.br/600x700/365/365010/produto/13341816/fe9a59131b.jpg",
      }
    }
    style={
      Object {
        "backgroundColor": "rgba(225, 226, 225, 1.00)",
        "borderColor": "rgba(225, 226, 225, 1.00)",
        "borderWidth": 2,
        "flex": 2,
        "height": 110,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 3,
        "flexDirection": "column",
        "justifyContent": "space-between",
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#333",
          "flex": 1,
          "fontSize": 15,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Vinho Nederburg Ros√©
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-around",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgba(114, 171, 65, 1.00)",
            "flex": 1,
            "fontSize": 16,
          }
        }
      >
        R$ 200,00
      </Text>
      <Button
        iconName="add-shopping-cart"
        onPress={[Function]}
        otherProps={Object {}}
        text="Add to Cart"
        textColor="#fff"
        withIcon={true}
      />
    </View>
  </View>
</TouchableOpacity>
`;
